# Confident Legal Research Application

- Main stakeholder/consumer: Junior Legal Assistant
- Distributed: No


**Goal:Improve quality & speed of case research for junior assistants**

## 1. Baseline System

To create an evaluation process we would need to have some baseline system in place. That is why we are creating a **Naive implementation** with the following initial setup:

- One dense vector embedding (`BAAI/bge-small-en-v1.5`, 384-dim, 512x512 context window)
- Search function with no reranker
- Naive RAG pipeline

In [2]:
%load_ext autoreload
%autoreload 2

### 1.1 Initialize dataset

We will use the famous CLERC dataset for this tutorial. Two reasons:
1. There exist queries in the dataset that we can use as a baseline for complex law queries 
2. The combination of queries/positive/negative responses provides a good baseline for the *golden set* that could be used for testing

In [3]:
from src.dataset import load_clerc


clerc_data = load_clerc("train", limit=2000)

clerc_data

Loading dataset shards:   0%|          | 0/31 [00:00<?, ?it/s]

Dataset to pydantic transformation:   0%|          | 0/2000 [00:00<?, ?it/s]

ClercSplit(corpus=41167 docs, queries=2000, qrels=2000)

In [4]:
sample_tuple = clerc_data.sample()

sample_tuple

(Query(query_id='101886', query='rights were created by contract; rather, the court merely held that the plaintiffs in that particular case had failed to demonstrate the existence of a compensable property right under their contract. Cf. Yankee Atomic Elec. Co. v. United States, 112 F.3d 1569, 1580 n. 8 (Fed.Cir.1997) (rejecting a nuclear utility’s argument that an annual assessment for the decontamination and decommissioning of uranium enrichment facilities effected a taking of the utility’s contract to purchase uranium enrichment services from the government at a specified price because the contract at issue did not contain an unmistakable promise that the utility would be immune from future assessments). In fact, as discussed in section II.C below, the Federal Circuit expressly held in  REDACTED  that a plaintiff may plead, in the alternative, both a breach of contract claim and a takings claim in the same complaint. Before the Federal Circuit’s decision in Stockton East II, there h

In [5]:
dict(list(clerc_data.qrels.items())[:2])


{'247774': {'22437426'}, '736597': {'195263'}}

### 1.2 Initialize vectors 

- use `qdrant_client` for storing vectors
- use `fastembed` for the simplest embeddings generation pipelines


We will also store the embeddings locally and try to handle them in parallel.

>NB!: This notebook is optimized to run on macbook devices. Try to adapt embedding config to your specific hardware

In [6]:
from qdrant_client import QdrantClient
import os

qdrant_url = os.getenv('QDRANT_URL')
qdrant_api_key = os.getenv('QDRANT_API_KEY')

client = QdrantClient(url=qdrant_url, api_key=qdrant_api_key)

In [7]:
# constants
COLLECTION_NAME = 'lawyer_citations'
DENSE_MODEL = "BAAI/bge-small-en-v1.5"
SIZE = 384

In [8]:
from src.indexing import EmbeddingConfig
from qdrant_client.models import Distance


config = EmbeddingConfig(
    name='dense_base',
    model_id=DENSE_MODEL,
    kind='dense',
    size=SIZE,
    distance=Distance.COSINE,
    providers=["CoreMLExecutionProvider", "CPUExecutionProvider"],
    parallel=None,
)

In [9]:
from src.indexing import EmbeddingCache
embedding_cache = EmbeddingCache()

In [10]:
from src.indexing import DocumentIndexer

document_vectors = DocumentIndexer(
    client,
    COLLECTION_NAME,
    embeddings=[config],
    cache=embedding_cache
)

In [11]:
document_vectors.ensure_collection()

In [20]:
document_vectors.upload(items=list(clerc_data.corpus.values()), batch_size=128)

embed:dense_base:   0%|          | 0/41167 [00:00<?, ?it/s]

In [12]:
client.get_collection(COLLECTION_NAME).points_count, (client.get_collection(COLLECTION_NAME).points_count or 0) > 0

(41167, True)

### 1.3 Create simple semantic search and RAG pipelines

- `DenseSearcher` is the simplest implementation of the semantic search: no sparse vectors, multivectors, no reranking. Just query API
- `LegalAssistant` - uses `DenseSearcher` for the retrieval and generates the results based on that. We use `anthropic api` with `lite-llm`

In [13]:
from src.search import DenseSearcher
from src.rag import LegalAssistant

qa = clerc_data.sample()
searcher = DenseSearcher(client, COLLECTION_NAME, config)
legal_assistant = LegalAssistant(searcher)


In [34]:
result = searcher.search(qa[0].query)
result[0].text, qa[1].text

('S.Ct. 1367, 4 L.Ed.2d 1435, was concerned with a due process argument relating to classifications established by nonentitlement provisions of the Act affecting aliens. The Court said, Ibid, at 611, 80 S.Ct. at 1373: “Particularly when we deal with a withholding of a noncontractual benefit under a social welfare program such as this, we must recognize that the Due Process Clause can be thought to interpose a bar only if the statute manifests a patently arbitrary classification, utterly lacking in rational justification.” Under the rubric of equal protection, Richardson, 404 U.S. at 81, 92 S.Ct. at 257, 30 L.Ed.2d 231, cites with approval Dandridge v. Williams, 397 U.S. 471, 487, 90 S.Ct. 1153, 25 L.Ed.2d 491, and says that a statutory classification in the area of social welfare is consistent with the Equal Protection Clause of the Fourteenth Amendment if it is “rationally based and free from invidious discrimination.” The Court goes on to say that: “While the present case, involving 

### 1.4 Simple test of the QA tool

In [14]:
answer = legal_assistant.ask("My clients landlord wants him to pay for the wall that was broken. No hand-in protocol was signed. Give me cases with similar problems")

print(answer)

Based solely on the case excerpts provided, none of the retrieved cases directly address the situation you have described — namely, a landlord seeking payment for a damaged wall where no hand-in (move-out inspection) protocol was signed. The excerpts cover the following unrelated legal topics:

1. Doc 9457344: A bankruptcy/debt dispute involving furniture repossession and post-petition claims — not related to property damage or move-out protocols.
2. Doc 14718586: A case about agricultural diversion payments and a tenant's obligation to assign payments to a landlord — not related to property damage or inspection protocols.
3. Doc 8705202: A government contract dispute involving equitable adjustments and administrative finality — not related to landlord-tenant property damage.
4. Doc 3706151: A landlord-tenant case under the Emergency Price Control Act of 1942 concerning rent overcharges — not related to property damage or move-out inspections.
5. Doc 4359176: A bankruptcy/receivership 

### 1.5 Baseline - LGTM dataset 

We can always start with the simplest approach, to understand if we even need to delve deep into the evaluations - a good old **Looks Good to Me (LGTM later)**. Stated simply: we can use either LLM or human to define if the queries are actually making good or no. This is a good starting point in case when decision whether system is working according to requirements or not.

In [15]:
lgtm_dataset = [
    'The plaintiff seeks expectation damages after the defendant failed to deliver goods under a commercial supply agreement. What precedents discuss breach of contract, expectation damages, and the duty to mitigate losses?', # Contract Law
    'A customer was injured on business premises after multiple intervening events contributed to the accident. What cases analyze duty of care, proximate cause, and foreseeability?', # Tort Law
    "My clients landlord wants him to pay for the wall that was broken. No hand-in protocol was signed. Give me cases with similar problems", # Tenant Law
]

#### Contract Law

In [16]:
answer = legal_assistant.ask(lgtm_dataset[0])

print(answer)

The excerpts address several relevant legal principles concerning breach of contract, expectation damages, and the duty to mitigate. Here is a summary of the key precedents and rules:

1. Proof of Damages and Reasonable Certainty

Under New York law, once it is established that a breach of contract has caused damages, the plaintiff need only prove the amount of those damages with "reasonable certainty" — not absolute precision. The foundational rule, articulated in Wakeman v. Wheeler & Wilson Mfg. Co., 101 N.Y. 205, 209 (1886), is that "[w]hen it is certain that damages have been caused by a breach of contract, and the only uncertainty is as to their amount, there can rarely be good reason for refusing, on account of such uncertainty, any damages whatever for the breach." Courts have further held that "[a] person violating his contract should not be permitted entirely to escape liability because the amount of the damages which he has caused is uncertain," and that "the wrongdoer must s

### Tort Law 


In [17]:
answer = legal_assistant.ask(lgtm_dataset[1])

print(answer)

Based on the provided excerpts, the following analysis applies to duty of care, proximate cause, and foreseeability in the context of customer injuries on business premises:

1. Duty of Care: Tort liability fundamentally depends on the violation of a duty of care owed to the injured person. The scope of that duty — whether it extends to bystanders, customers, or other classes of plaintiffs — is ordinarily defined by the common law. Legislatures can create new duties of care, but the mere fact that a statute defines due care does not, by itself, create a tort-enforceable duty. Courts have consistently held that a plaintiff must first establish that the defendant owed them a duty before any further negligence analysis can proceed.

2. Proximate Cause: Causation is described as "the most susceptible to summary determination" among all elements necessary to support recovery in a tort action. To demonstrate proximate cause, a plaintiff must show that the defendant, through affirmative actio

In [18]:
answer = legal_assistant.ask(lgtm_dataset[2])

print(answer)

Based solely on the case excerpts provided, none of the retrieved cases directly address the specific legal issue of a landlord seeking payment for property damage (a broken wall) in the absence of a signed hand-in (move-out/check-out) protocol. The excerpts cover unrelated legal matters, including bankruptcy and furniture repossession (doc_id: 9457344), agricultural payment assignments and landlord rent notes (doc_id: 14718586), government contract equitable adjustments (doc_id: 8705202), emergency rent control regulations and overcharges (doc_id: 3706151), and bankruptcy bar orders and asset distribution (doc_id: 4359176). None of these cases involve a landlord's claim for property damage to a rental unit, nor do they address the evidentiary or contractual significance of a missing hand-in or move-out protocol.

Therefore, the provided excerpts do not contain sufficient information to answer your question about cases involving a landlord's claim for wall damage in the absence of a si

### 1.6 Results interpetation 

From the first side the AI generated queries are being passed. However, if you ask stronger LLM model to analyze the actual result it will identify two issues with the results if the `lgtm_dataset[0]` and `lgtm_dataset[1]` queries replies:

**Confidence exceeds evidence** (prompt discipline is broken - section 3 and summary are full of speculation in first query) - grave mistake cause it will confuse the Junior Assistants in the same vein it does us. Law needs not just confidence, but a lot of precision

That is why I urge the practitioners to write some queries without using AI if they want to judge manually (AI generated queries -> stronger AI judgements, human generated queries -> both human and AI judgements can work). For the `lgtm_dataset[2]` we see a clear outlier: `Based on the case excerpts provided, none of them directly address the specific legal issue `. This either means that our query is wrong for this dataset (mostly this question was from personal experience in EU not US), however this does not reflect the result => our naive implementation of search still gave wrong results. This is a clear signal now that evaluation is needed to setup a **baseline/AS IS**

## 2 Basis of the evaluation framework

Even on a single query, the retrieval layer is visibly underperforming — by design. The PoC intentionally omits sparse vectors, reranking, and query rewriting so the failure modes those techniques usually hide surface up front.

What saved this run is the LLM: `claude-sonnet-4.6` is strong enough to notice the retrieved excerpts don't answer the question and refuse to fabricate. That safety net is fragile. With a weaker model, an opaque pipeline, or domain experts who can't audit the citation chain, the same retrieval bug becomes invisible — and the system confidently misleads a junior associate who can't yet tell the difference.

### 2.1 Goal oriented evaluation specification

The evaluation is designed **top-down**: a system that doesn't serve its business goal isn't worth measuring at the component level.

### 2.2 Inputs/Assumptions

The preconditions feeding into the methodology:
- **GOAL:** Confident Legal Research
- **CONSUMER:** Junior associate — limited domain depth, needs guidance and explanations
- **DISTRIBUTED:** No — single goal, single consumer

In [33]:
from IPython.display import HTML

HTML("""
<div style="text-align: center;">
    <img src="diagrams/eval-inputs.svg" width="300">
</div>
""")

### 2.3 Goal Decomposition

**Goal decomposition** bridges the abstract business goal and the concrete metrics that measure it. Before we can quantify, we have to be specific about *what the system is* and *what it must do*.

Two questions to anchor the decomposition:
- **Is this a *business* goal?** Yes — it describes the system's *utility*, not its internals.
- **Can we make *component-level* assumptions?** Yes — the system has two: **searcher** (retrieval, with well-understood mechanics) and **rag** (LLM-driven generation).

Refined business objective:

`Improve quality & speed of case research for junior associates`

Technical commitments already baked into the codebase:

`Confident & fast legal agentic search` *(too vague to test as-is — decomposition follows)*

Decomposition is one of the oldest design principles in software engineering. Here we use it to turn the vague tech goal into the evaluation context:

![System Goal Decomposition](diagrams/system-goal-decomposition.svg)

At the system level, three NFRs apply to every sub-system:
- **Trustworthiness** — can the consumer rely on the output without verification, and fail loudly when it can't?
- **Correctness** — does the output match ground truth for the task it owns?
- **Performance** — is it fast enough to be useful in the consumer's loop?

These stay vague until we map them to individual components. Each sub-system has a different consumer, and the consumer determines what each NFR *operationally means*:

![Component Level Goal Decomposition](diagrams/component-level-goal-decomposition.svg)

**Search** is consumed by RAG (programmatic). Its failure surface is narrow — a bad ranked list, an over-confident empty result, or a slow query. Three flat axes are enough.

**RAG** is consumed by a junior associate (human). Its failure surface is richer — confidently-wrong differs sharply from correctly-refused, and citation grounding, hedging, and clarity are all separately damaging when they fail. So the decomposition goes deeper on the RAG side.

**The depth asymmetry is itself the finding:** it tells the reader where the evaluation effort has to concentrate.

One consequence worth calling out early: dense semantic search has no natural zero — every query embeds *somewhere*, every doc embeds *somewhere*, cosine similarity always produces an ordering. So *no relevant doc in the corpus → empty top-K* isn't emergent behavior; it has to be engineered via a calibrated threshold. That's why **Calibration** sits as the headline trustworthiness bullet on the search side.

Without per-component decomposition, a wrong final answer can't be localized to bad retrieval vs. bad synthesis — and what can't be localized can't be fixed. That's exactly what the current PoC shows: RAG had to reject its own retriever's ranking because there was no earlier signal — no calibrated empty, no per-component metric — that would have told the operator the search side was the problem.

## 3. Direct Mapping of the Non-functional requirement into the metrics

### 3.1 Ranking Quality Mapping


Next part of the tutorial we will focus only on one non-functional requirement that obviously needs some improvement by the simple implementation we have. I am talking about **Ranking quality**, a search component NFR (non-functional requirement, will be used from now on).

What we do is walk through a series of decision questions that help identify the most appropriate metric to use in this situation.

![Evaluation Metric Definition](./diagrams/evaluation-path-to-metric.svg)


The end result is that currently have a clear path why we would measure `NDCG@10` - because:
- We have the golden dataset from `CLERC`
- We want to measure what is the current **Ranking Quality** of the system

We simplified two routes here though: 
- There is no validation of the coverage / leakage / drift for the dataset, that is a separate topic within itself
- We assume that for now it will be enough to ignore the consumer being an LLM instead of a person. Technically this can be a mistake — however, acknowledging that our metric is flawed and doing the best we can with it is better than not measuring at all.

Let us run `NDCG@10` against our dataset and interpret the results

In [19]:
from src.evaluation import evaluate_retrieval

report = evaluate_retrieval(searcher, clerc_data)  # smoke run
print(report)

search:   0%|          | 0/2000 [00:00<?, ?it/s]

Retrieval evaluation (2000 queries, top-10):
  ndcg@10      0.1384
  recall@10    0.2890
  mrr@10       0.0931


Our `ndcg@10` on a 2000-query slice of CLERC **train** is `0.1384`. This lands in the same neighborhood as the published LegalBERT-DPR result of `0.147` — see other [benchmarks](https://www.emergentmind.com/topics/clerc-dataset).

**Caveat:** the published number is measured on the full CLERC *test* split; we're on a subset of *train*. Treat this as a directional sanity check ("we're in the right ballpark for an off-the-shelf dense encoder on legal text"), not a benchmark reproduction. To make the comparison honest, we'd need to rerun on the full test split.

Two reasons the number is low in absolute terms:
1. The legal dataset is domain specific, highly lexical and has complex queries
2. The dataset we are using is actually a trimmed version of the real one. We might expect lower results on the full corpora

## 4.0 Improvement strategies

### 4.1 Trying to naively use Qdrant skills


We can now use qdrant best practices and ask our LLM to use claude skills to improve the system. Its as easy as prompting into the claude-code the following:
```
Use qdrant skills: https://skills.qdrant.tech/search?query=about%20search%20quality%20improvements to find how to improve the current naive search from @src/search.py
```

![Claude Code Running](./diagrams/improvement-claude-skills.png)




Most probably it will go into one of the following strategies:
- [Use Hybrid Search with BM25](https://qdrant.tech/documentation/search/hybrid-queries/?selector=aHRtbCA%2BIGJvZHkgPiBtYWluID4gc2VjdGlvbiA%2BIGRpdiA%2BIGRpdiA%2BIGRpdjpudGgtb2YtdHlwZSgyKSA%2BIGRpdiA%2BIGRpdjpudGgtb2YtdHlwZSgxKSA%2BIGFydGljbGUgPiBoMjpudGgtb2YtdHlwZSgxKQ%3D%3D&q=Hybrid+Search)
- [Use richer embeddings](https://qdrant.tech/documentation/manage-data/points/)

### 4.2 Hybrid Search as a solution

This time it generated for me Hybrid Search class. Since we plan to also create a new collection we can increase the dense embeddings size as well  

In [ ]:
import pandas as pd 
from fastembed import TextEmbedding

supported_models = (
    pd.DataFrame(TextEmbedding.list_supported_models())
    .sort_values("size_in_GB")
    .drop(columns=["sources", "model_file", "additional_files"])
    .reset_index(drop=True)
)
supported_models

In [15]:

DENSE_MODEL_V2  = 'BAAI/bge-base-en-v1.5' #  0.320 GB,	768 DIM	
DENSE_SIZE_V2 = 768
COLLECTION_NAME_V2 = 'trec_dl_v2'
SPARSE_MODEL_V2 = 'Qdrant/bm25' 

In [16]:
from src.indexing import EmbeddingConfig
from qdrant_client.models import Distance
from qdrant_client.models import Modifier


configs_v2 = [
    EmbeddingConfig(
        name='dense_base',
        model_id=DENSE_MODEL_V2,
        kind='dense',
        backend='sentence-transformers',
        size=DENSE_SIZE_V2,
        distance=Distance.COSINE,
    ),
    EmbeddingConfig(
        name="bm25",
        model_id=SPARSE_MODEL_V2,
        kind="sparse",
        modifier=Modifier.IDF
    )
]

In [17]:
from src.indexing import DocumentIndexer

document_indexer_v2 = DocumentIndexer(
    client,
    COLLECTION_NAME_V2,
    embeddings=configs_v2,
    cache=embedding_cache
)

In [18]:
document_indexer_v2.ensure_collection()

In [19]:
document_indexer_v2.upload(items=list(clerc_data.corpus.values()), batch_size=256)

In [19]:
from src.search import HybridSearcher

searcher_v2 = HybridSearcher(
    client, 
    COLLECTION_NAME_V2, 
    configs_v2,
)

In [20]:
qa_v2 = clerc_data.sample()

qa_v2

(Query(query_id='623425', query='the constitutional rights of nonparty media organizations. IV FREEDOM OF SPEECH A. The Appropriate Legal Standard The petitioners contend that the district court’s order is a prior restraint on their first amendment right to free speech. We agree that the district court’s order is properly characterized as a prior restraint. See M. Nimmer, supra, § 4.03. Prior restraints are subject to strict scrutiny because of the peculiar dangers presented by such restraints. See id. § 4.04; L. Tribe, American Constitutional Law § 12-32 (1978). Accordingly, the district court’s order may be upheld only if the government establishes that: (1) the activity restrained poses either a clear and present danger or a serious and imminent threat to a protected competing interest,  REDACTED see Wood v. Georgia, 370 U.S. 375, 383-85, 82 S.Ct. 1364, 1369-70, 8 L.Ed.2d 569 (1962); (2) the order is narrowly drawn, Carroll v. President and Commissioners of Princess Anne, 393 U.S. 1

In [21]:
result = searcher_v2.search(qa_v2[0].query)
result[0].text, qa_v2[1].text

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

Fetching 18 files:   0%|          | 0/18 [00:00<?, ?it/s]

('§ 3531.9 (1984). The rule against third party standing is subject to several well-established exceptions. See id. Under the present procedural posture of this case, none of those exceptions apply. Accordingly, we conclude that the petitioners lack standing to assert the constitutional rights of nonparty media organizations. IV FREEDOM OF SPEECH A. The Appropriate Legal Standard The petitioners contend that the district court’s order is a prior restraint on their first amendment right to free speech. We agree that the district court’s order is properly characterized as a prior restraint. See M. Nimmer, supra, § 4.03. Prior restraints are subject to strict scrutiny because of the peculiar dangers presented by such restraints. See id. § 4.04; L. Tribe, American Constitutional Law § 12-32 (1978). Accordingly, the district court’s order may be upheld only if the government establishes that: (1) the activity restrained poses either a clear and present danger or a serious and imminent threa

In [16]:
from src.evaluation import evaluate_retrieval

report_v2 = evaluate_retrieval(searcher_v2, clerc_data)  # smoke run
print(report_v2)

It is top 10 evaluation


search:   0%|          | 0/2000 [00:00<?, ?it/s]

Retrieval evaluation (2000 queries, top-10):
  ndcg@10      0.1760
  recall@10    0.3360
  mrr@10       0.1277


### 4.2 Hybrid Search analysis

This is where our theoretical framework stoppes working and reality kicks in. 

We will notice many more times that there are cases when one metric is not enough to tell the whole picture, even if we mapped it one to one in the methodology. If we think carefull about our NFR is to have **good ranking quality**. So far we have still do not know how good is *ndcg@10* - it became better but its still much to be desired. 


This means that there is something else we might have missed. This requires us to check other metrices. Always prefer standard metrices to which you know what it represents - we are flipping the script here and going bottom-up. Thus the most default metrices we can have at this level:
- recall@10
- mrr@10 

Notice that recall@10 is way higher than the ndcg@10 which means we can create a hypothesis: ranking is suboptimal, not the results. 

To check it more, lets get bigger sample of `metric@100`

In [14]:
from src.evaluation import evaluate_retrieval, RetrievalMetric

report_v2_100 = evaluate_retrieval(
    searcher_v2, 
    clerc_data,
    top_k=100,
    metrics=[
        RetrievalMetric.RECALL_100,
        RetrievalMetric.NDCG_100, 
        RetrievalMetric.MRR_100,
    ],
)  # smoke run
print(report_v2_100)

search:   0%|          | 0/2000 [00:00<?, ?it/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

Retrieval evaluation (2000 queries, top-100):
  recall@100   0.9765
  ndcg@100     0.3120
  mrr@100      0.1546


### 4.3 Statistical significance

Lets check now how relatively well does our new solutions? Using naive approach we get the following:

In [24]:
ndcg_10_naive =  0.1384
ndcg_10_hybrid = 0.1760
percent = (1 - (ndcg_10_naive/ndcg_10_hybrid)) * 100

f'{percent:0.2f}%'

'21.36%'

Impressive isn't it? As with everything in computer science we have to go deeper and ask: is it real though? 

Let's say that now because we have a new embedding, we get a certified win for just one query where NDCG goes from 0.05 to 0.95. However altogether we have thousands of queries. The aggregate still shifts up - but attributing that shift to system change is wrong. Aggregate-only reports can't distinguish many small real improvements from one lucky query.   


So next what we are doing is called `paired testing` which allows us to see the real significance change. The claim is that because each query is scored under both systems that means they are comparable as pairs.  

##### Deep dive into the difference between two results

 How much better, with what confidence, and at what significance `HybridSearch` beast `DenseSearch` for 2000 queries? This test is formally called [Wilcoxon](https://en.wikipedia.org/wiki/Wilcoxon_signed-rank_test)

In [28]:
from src.evaluation import compare_pair, RetrievalMetric

report = compare_pair(
    { 'dense': searcher, 'hybrid': searcher_v2 },
    data=clerc_data,
    metrics=[RetrievalMetric.NDCG_10, RetrievalMetric.RECALL_10, RetrievalMetric.MRR_10],
    top_k=10
)

report

search:dense:   0%|          | 0/2000 [00:00<?, ?it/s]

search:hybrid:   0%|          | 0/2000 [00:00<?, ?it/s]

PairwiseComparisonReport(name_a='dense', name_b='hybrid', num_queries=2000, top_k=10, ci_level=0.95, n_bootstrap=10000, n_permutations=10000, metrics=[PairwiseMetric(metric=<RetrievalMetric.NDCG_10: 'ndcg@10'>, mean_a=0.13840649052670598, mean_b=0.17548984056145725, delta=0.037083350034751286, ci_low=0.026105270374789064, ci_high=0.04817672301692784, p_value=4.3022567078467084e-10, win_rate=0.229), PairwiseMetric(metric=<RetrievalMetric.RECALL_10: 'recall@10'>, mean_a=0.289, mean_b=0.337, delta=0.048, ci_low=0.027, ci_high=0.0685, p_value=7.957499027053408e-06, win_rate=0.1395), PairwiseMetric(metric=<RetrievalMetric.MRR_10: 'mrr@10'>, mean_a=0.09312440476190476, mean_b=0.12677718253968256, delta=0.03365277777777778, ci_low=0.02356799603174603, ci_high=0.04356901289682539, p_value=2.8538014386785227e-09, win_rate=0.229)])

In [30]:
print(report)

hybrid vs dense (2000 queries, top-10, 95% CI):
  NDCG@10:    Δ = +0.037   95% CI [+0.026, +0.048]   p = 4.3e-10   hybrid>dense in 22.9% of queries
  Recall@10:  Δ = +0.048   95% CI [+0.027, +0.069]   p = 8.0e-06   hybrid>dense in 14.0% of queries
  MRR@10:     Δ = +0.034   95% CI [+0.024, +0.044]   p = 2.9e-09   hybrid>dense in 22.9% of queries


#### Reading the numbers 

The improvement is real. The three headline numbers say so:
- **Δ = +0.037** — mean NDCG@10 improvement per query. This is the primary quantity.
- **95% CI [+0.026, +0.048]** — tight, does not cross zero. Not a fluke.
- **p = 4.3e-10** — at N=2000, significance is *cheap*: any real effect gives tiny p-values.
  Significance is a *floor* here, not a *win*.


**Hybrid > Dense on only 22.9% of queries.**

Two consequences:
- Aggregates result of delta hides where the gain lives
- A cascade or router approach might beat either system alone - if we can predict where hybrid wins we could improve the overall performance

### 4.4 Coming up to the improvement framework 

Formally we already got the result, and came up naturally to another looped methodology which we call integration layer (since it can be easily integrated into out Continous Integration pipeline to serve as both safeguard and addressable goal). Here is a more detailed generalized version of the improvement loop we are currently in: 


![From Evaluation to Improvements](./diagrams/from-evaluation-to-integration.svg)

### 4.5 Do we want to improve more?

§4.3 answered is the delta real? — the inner-loop gate of the diagram above. §4.5 asks a different question: is the current state good enough, or do we open another improvement session? That's the outer-loop gate, and it lives outside every metric on the dashboard.

Sometimes and this is this time the simple solutions really work. However, we should be aware that for the next problem of having a 
disperancy between `Hybrid:NDCG@10` and `Hybrid:NDCG@100` in order of almost 80% - we cannot fully rely on the solutions that AI will give us. LLMs will just take the easiest route of selecting the **reranking** strategy. Now, according to the common consensus - reranking is the most powerful tool (recovering up to [40% of precision](https://app.ailog.fr/en/blog/guides/reranking)) to use for improving search. 

Two branches from the satisfaction gate:

- Satisfied → move to the next NFR or the next component.
- Not satisfied → write a new improvement objective, not a new technique. Rerun the loop of improvement checking on the new model:

For this article, we're not satisfied — for a specific reason. Hybrid:NDCG@10 and Hybrid:NDCG@100 diverge by ~80%: the relevant document is often in the top-100 but not in the top-10. That's a ranking problem, not a recall problem, and it gives us a concrete objective — close the @10/@100 gap — before we name any technique.

An objective thus becomes: reach acceptable (>60% NDCG) based on the NDCG@100 being close to 99% -> capacity is there, ranking performance needs to be worked on.

The hypotheses split cleanly across two subcomponents:
- Representation side 
- Retrieval side 

#### Representation side 
Either the vector space itself isn't discriminative enough at the top of the ranking, or the encoder never sees enough of the passage to build a rich vector in the first place. Two hypotheses live here:
- Chunking — passages exceed the 512-token context of bge-base; the encoder is silently truncating, so more space would give it room to capture the full complexity of the corpus.
- Legal-tuned encoder — a domain-scoped model captures features (statute references, citation patterns, legal-specific semantics) that a general-purpose encoder flattens.

#### Retrieval side 

The representation is fine; what's broken is the ordering given the representation. The candidate here is reranking, and it further splits by architecture:

- Cross-encoder — full query-document attention, highest precision, slowest.
- Late-interaction (e.g. ColBERT) — token-level scoring at index time, faster than cross-encoder, richer than a bi-encoder.


#### Chunking 

Let's start with the hypothesis that `CLERC` has huge legal texts that small encoders cannot simply represent it well. For this will we analyze text corpora (the one we have) if it actually can be tokenized by the `BGE` encoder  

In [32]:
from src.corpus_analysis import analyze_corpus

stats = analyze_corpus(list(clerc_data.corpus.values()))

print(stats)

tokenize:   0%|          | 0/42 [00:00<?, ?it/s]

Corpus stats (41,167 docs, tokenizer=BAAI/bge-base-en-v1.5):
  Total tokens:      21,608,613
  Mean tokens/doc:   524.9
  Length quantiles:  p50=521  p90=600  p95=627  p99=706  max=1520
  Over model max (512 tok): 55.7% of docs
  Truncation info loss: 5.7% of total tokens discarded
  Length distribution:
    >  128 tok: 100.0%
    >  256 tok: 100.0%
    >  512 tok:  55.7%
    > 1024 tok:   0.0%
    > 2048 tok:   0.0%


In [33]:
if stats.needs_chunking():
    print('-> chunking recommended')

-> chunking recommended


Even though the code says that chunking is needed, we always need to analyze the signals first and foremost. The stats themselves give us clearer picture: only 5.7% of total tokens are discarded. Not worth investing time. Especially since each addition to the initial solution comes with certain tradeoffs. If on the other side of the tradeoff we do not get any benefit, we should better skip. In case of chunking for this case study, it comes with:

##### Downsides
- Optimal chunking is contextual, you have to split based on some assumption about the text - requires deep dive into corpora
- Chunking can increase disk space `2x` or even `3x` depending on chunking strategy
- Chunking causes duplication of the results -> means harder query result filtering and post processing

##### Upsides (in this case study)
- 5% improvement in text representation


Clearly a pass

#### Finetuned encoder model

We are testing our assumption that bigger and more specialized encoder could actually improve our system and NDCG metric.  

Let us use [fine-tuned modernbert model](https://free.law/2025/03/11/semantic-search/) available on [hugging face](https://huggingface.co/freelawproject/modernbert-embed-base_finetune_512).


We are doing same steps as before: initializing new collection and adding new embeddings to it. 

In [22]:
from src.indexing import EmbeddingConfig
from qdrant_client.models import Distance
from qdrant_client.models import Modifier


DENSE_MODEL_V3  = 'Free-Law-Project/modernbert-embed-base_finetune_512' #  0.320 GB,	768 DIM	
DENSE_SIZE_V3 = 768
COLLECTION_NAME_V3 = 'trec_dl_v3'
SPARSE_MODEL_V3 = 'Qdrant/bm25' 


configs_v3 = [
    EmbeddingConfig(
        name='dense_base',
        model_id=DENSE_MODEL_V3,
        kind='dense',
        backend='sentence-transformers',
        size=DENSE_SIZE_V3,
        distance=Distance.COSINE,
        device='mps'
    ),
    EmbeddingConfig(
        name="bm25",
        model_id=SPARSE_MODEL_V3,
        kind="sparse",
        modifier=Modifier.IDF
    )
]

In [23]:
from src.indexing import DocumentIndexer

document_indexer_v3 = DocumentIndexer(
    client,
    COLLECTION_NAME_V3,
    embeddings=configs_v3,
    cache=embedding_cache
)

document_indexer_v3

In [24]:
document_indexer_v3.ensure_collection()

In [18]:
document_indexer_v3.upload(items=list(clerc_data.corpus.values()), batch_size=32)

Loading weights:   0%|          | 0/134 [00:00<?, ?it/s]

Batches:   0%|          | 0/1287 [00:00<?, ?it/s]

Evaluate new indexer

In [25]:
from src.search import HybridSearcher

searcher_v3 = HybridSearcher(
    client, 
    COLLECTION_NAME_V3, 
    configs_v3,
)

In [22]:
from src.evaluation import evaluate_retrieval, RetrievalMetric

report_v3_10 = evaluate_retrieval(
    searcher_v3, 
    clerc_data,
    top_k=10,
    metrics=[
        RetrievalMetric.RECALL_10,
        RetrievalMetric.NDCG_10, 
        RetrievalMetric.MRR_10,
    ],
)  # smoke run
print(report_v3_10)

search:   0%|          | 0/2000 [00:00<?, ?it/s]

Loading weights:   0%|          | 0/134 [00:00<?, ?it/s]

Retrieval evaluation (2000 queries, top-10):
  recall@10    0.3685
  ndcg@10      0.1924
  mrr@10       0.1394


Overall the results are the following: 

| Metric    | Naive (bge-small) | Hybrid (bge-base + BM25) | Legal ModernBERT |
| --------- | ----------------: | -----------------------: | ---------------: |
| ndcg@10   |            0.1384 |                   0.1755 |           0.1924 |
| recall@10 |            0.2890 |                   0.3365 |           0.3685 |
| mrr@10    |            0.0931 |                   0.1269 |           0.1394 |


We can see gradual improvement in the results for `Legal ModernBERT` of around ~10%. Its good and signifies to us that we are on the right track however, not nearly enough to justify the slowness that `Legal ModernBERT` brings with it retrieval (on Macbook machines at least). 

Thus, we can try to address the elephant in the room by the name of `reranker`. 

In [ ]:
from fastembed.rerank.cross_encoder import TextCrossEncoder

TextCrossEncoder.list_supported_models()

In [26]:
from src.rerank import Reranker, CrossEncoderStrategy


searcher_v4 = Reranker(
    retriever=searcher_v2,
    strategy=CrossEncoderStrategy(device='mps', model_id="Xenova/ms-marco-MiniLM-L-6-v2"),
    candidates=100
)

In [51]:
res = searcher_v4.search(qa[0].query)

res, qa[0].query

([SearchHit(doc_id='23710569', text='may not terminate benefits absent a showing of previous clear and specific error or medical improvement which is sufficient to establish that an applicant is no longer “continuously disabled as so defined.” B. Case Law Although it is scarce and by no means (in itself) dispositive on the issue, the case law in the area lends support to Finnegan’s interpretation of the grandfather clause. While no case directly considers the question of terminating SSI benefits to former state recipients who are covered by the. grandfather clause, several cases focusing principally on the rights of “rollbacks”, see note 1, supra, discuss the position of grandfatherees in an effort to distinguish the eligibility of the two groups. The dicta in each of these cases indicate that their judicial authors considered grandfatherees to be automatically eligible for SSI benefits and exempt from HEW’s “initial determinations.” The Courts of Appeals have consistently referred to 

In [ ]:
from src.evaluation import evaluate_retrieval

report_v4 = evaluate_retrieval(searcher_v4, clerc_data)
print(report_v4)

In [53]:
report_v4 = evaluate_retrieval(searcher_v4, clerc_data, limit=10)
print(report_v4)

search:   0%|          | 0/10 [00:00<?, ?it/s]

Retrieval evaluation (10 queries, top-10):
  ndcg@10      0.0946
  recall@10    0.2000
  mrr@10       0.0625


In [41]:
report_v5 = evaluate_retrieval(searcher_v2, clerc_data, limit=10)
print(report_v5)

search:   0%|          | 0/10 [00:00<?, ?it/s]

Retrieval evaluation (10 queries, top-10):
  ndcg@10      0.0764
  recall@10    0.2000
  mrr@10       0.0393


#### Trying reranker with hybrid search from cloud inference 

In [29]:
inference_client = QdrantClient(
    url=os.getenv('QDRANT_INF_URL') or 'https://af3b00d6-3a14-4757-883d-08222884e98a.ap-southeast-2-0.aws.cloud.qdrant.io',
    api_key=os.getenv('QDRANT_INF_API_KEY') or 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhY2Nlc3MiOiJtIiwic3ViamVjdCI6ImFwaS1rZXk6Mzk2YTEwMDItZmZmZS00MmJiLTgzNDItMDEwODI2NzYxMjVlIn0.i8DVj4bJSo2GU1VO5A5aYS06Bu17Mboj2C3-x8D5o0c',
    cloud_inference=True,
    timeout=300,
)

In [30]:
COLLECTION_NAME_V5 = 'clerc_v5'
configs_v5 = [
    EmbeddingConfig(
        name='dense',   
        model_id='mixedbread-ai/mxbai-embed-large-v1',
        kind='dense',
        size=1024, 
        distance=Distance.COSINE
    ),
    EmbeddingConfig(
        name='splade',
        model_id='prithivida/splade_pp_en_v1',
        kind='sparse',
        modifier=Modifier.IDF
    ),
    EmbeddingConfig(
        name='colbert', 
        model_id='answerdotai/answerai-colbert-small-v1', 
        kind='late_interaction', 
        size=96, 
        distance=Distance.COSINE
    ),
]

In [31]:
inference_client.get_collections()

CollectionsResponse(collections=[CollectionDescription(name='clerc_v5'), CollectionDescription(name='gooaq')])

In [32]:
indexer_v5 = DocumentIndexer(inference_client, COLLECTION_NAME_V5, configs_v5)

In [33]:
indexer_v5.ensure_collection()

In [36]:
from qdrant_client import models

inference_client.update_collection(
    "clerc_v5",
    vectors_config={"colbert": models.VectorParamsDiff(hnsw_config=models.HnswConfigDiff(m=0))},
)

True

In [38]:
indexer_v5.upload(list(clerc_data.corpus.values()), 64, 5, True)

upsert:   0%|          | 0/492 [00:00<?, ?it/s]

In [ ]:
configs_v5[:2]

In [47]:
from src.search import HybridSearcher

searcher_v6 = HybridSearcher(
    inference_client, 
    COLLECTION_NAME_V5, 
    configs_v5[:2],
)

In [ ]:
searcher_v6.search("hello there")

In [62]:
report_v6 = evaluate_retrieval(searcher_v6, clerc_data, limit=100)
print(report_v6)

search:   0%|          | 0/100 [00:00<?, ?it/s]

Retrieval evaluation (100 queries, top-10):
  ndcg@10      0.1217
  recall@10    0.2400
  mrr@10       0.0849


In [51]:
from src.doc_mapping import load_pid2did, analyze_siblings

pid2did = load_pid2did(clerc_data.corpus)

In [ ]:
from src.evaluation import evaluate_retrieval_graded
report_v6 = evaluate_retrieval_graded(searcher_v6, clerc_data, pid2did, limit=100)
print(report_v6)

search:   0%|          | 0/100 [00:00<?, ?it/s]

Retrieval evaluation (100 queries, top-10):
  ndcg@10      0.1236
  recall@10    0.2262
  mrr@10       0.1081


In [ ]:
from src.search import HybridRerankSearcher
from src.rerank import Reranker

searcher_v7 = HybridRerankSearcher(
    inference_client,
    COLLECTION_NAME_V5,
    prefetch=[configs_v5[0], configs_v5[1]],   # dense + splade, fused with RRF
    rerank=configs_v5[2],                      # colbert rescores ~30-100 candidates only
    prefetch_multiplier=25,
)


rerank_searcher = Reranker(
    searcher_v7,
    strategy=CrossEncoderStrategy(device='mps', model_id="Xenova/ms-marco-MiniLM-L-6-v2")
)

Fetching 5 files:   0%|          | 0/5 [00:00<?, ?it/s]

In [64]:
from src.evaluation import evaluate_retrieval
report_v7 = evaluate_retrieval(rerank_searcher, clerc_data, limit=100)
print(report_v7)

search:   0%|          | 0/100 [00:00<?, ?it/s]

Retrieval evaluation (100 queries, top-10):
  ndcg@10      0.0923
  recall@10    0.2000
  mrr@10       0.0601


In [71]:
inference_client.delete_payload_index(COLLECTION_NAME_V5, "doc_id")

UpdateResult(operation_id=553, status=<UpdateStatus.COMPLETED: 'completed'>)

In [72]:
from qdrant_client.models import PayloadSchemaType

inference_client.create_payload_index(
    collection_name=COLLECTION_NAME_V5,
    field_name="doc_id",
    field_schema=PayloadSchemaType.KEYWORD,
)

UpdateResult(operation_id=555, status=<UpdateStatus.COMPLETED: 'completed'>)

In [76]:
from src.search import HybridRerankSearcher
from src.rerank import Reranker, QdrantLateInteractionStrategy


rerank_searcher_v2 = Reranker(
    searcher_v6,
    strategy=QdrantLateInteractionStrategy(
        client=inference_client,
        collection_name=COLLECTION_NAME_V5,
        vector_name=configs_v5[2].name,
        model_id=configs_v5[2].model_id
    ),
    candidates=100
)

In [ ]:
rerank_searcher_v2.search("hello")

In [79]:
from src.evaluation import evaluate_retrieval
report_v8 = evaluate_retrieval(rerank_searcher_v2, clerc_data, limit=10)
print(report_v8)

search:   0%|          | 0/10 [00:00<?, ?it/s]

Retrieval evaluation (10 queries, top-10):
  ndcg@10      0.1387
  recall@10    0.2000
  mrr@10       0.1200


In [83]:
from src.rerank import Reranker, CrossEncoderStrategy

rerank_searcher_bge = Reranker(
    searcher_v6,
    strategy=CrossEncoderStrategy(
        model_id="BAAI/bge-reranker-base",
        device="mps",
    ),
    candidates=120,
)

In [ ]:
report_bge = evaluate_retrieval(
    rerank_searcher_bge, clerc_data, limit=100,
)
print(report_bge)

search:   0%|          | 0/100 [00:00<?, ?it/s]

Even though we have tried everything the improvement gap is relatively small - the reranker is giving us about 20% of improvement. This however is still suboptimal solution for our purposes, and knowing that we have a subset of the real data (20K points) we might have some doubts if we are doing everything right? 

#### 4.6. Know your data!

What does our data tells us about itself? Lets check some things